### PR TITLE
ENG-13194  Fix failure to shutdown a node with@StopNode

### PR DESCRIPTION
**Pull request by zeemanhe**
When MP sites are in the middle of transaction processing, @StopNode  may cause MP deadlock and fail to stope a node. The pull request tries to terminate these transactions with a poison fragment response and then shutdown  MP Ro site pool.  and also offer a fake task to unblock sites to be shutdown.

### CODE REVIEW

File: src/frontend/org/voltdb/iv2/BaseInitiator.java [hunk 1]
```diff
@@ -199,6 +199,8 @@ public void shutdown()
 
         if (m_siteThread != null) {
             try {
+                //add a null task to unblock the queue
+                m_scheduler.getQueue().offer(Scheduler.m_nullTask);
```
Comment (luyangco) [inline src/frontend/org/voltdb/iv2/BaseInitiator.java]: SpScheduler already does that in `SpScheduler.shutdown()`,  you just need to offer a null task to MpScheduler.

MpScheduler owns `m_tasks` which can be used to offer null task.

File: src/frontend/org/voltdb/iv2/MpRoSite.java [hunk 2]
```diff
@@ -95,6 +95,8 @@
     // Current topology
     int m_partitionId;
 
+    private TransactionState m_txnState = null;
```
Comment (luyangco) [inline src/frontend/org/voltdb/iv2/MpRoSite.java]: Maybe add a comment to mention about what does it mean when `m_txnState` is not null

File: src/frontend/org/voltdb/iv2/MpRoSitePool.java [hunk 3]
```diff
@@ -221,7 +227,8 @@ boolean doWork(long txnId, TransactionTask task)
                             m_poolThreadFactory));
             }
             site = m_idleSites.pop();
-            m_busySites.put(txnId, site);
+            busySites.put(txnId, site);
+            m_busySites.set(busySites);
```
Comment (luyangco) [inline src/frontend/org/voltdb/iv2/MpRoSitePool.java]: Before we merge this, we should measure the read-only MP queries performance.

http://ci/job/performance-graphs/lastSuccessfulBuild/artifact/performance-graphs-90-days/stats-index.html#ProjectionMP-cached%202%20nodes

Comment (luyangco) [reply src/frontend/org/voltdb/iv2/MpRoSitePool.java]: What if while the MpScheduler is shutting down, a task reaches to MpRoSite, pop a site out from `m_idleSites` but before it put into `m_busySites`. Shutdown path could miss this one in both idle and busy sites, causing an incomplete shutdown.

I think the right way to do is to synchronize both here and shutdown path.

Comment (zeemanhe) [reply src/frontend/org/voltdb/iv2/MpRoSitePool.java]: MpRoSitePool will be locked down and accept no more work upon shutting down. So this should not be an issue.

**Comment by luyangco** [comment]
Please remember to run the read-only Mp performance test before merge. Otherwise it looks good to me.

**Pull request merged**